taichi/backends/device.cpp

```diff
@@ -21,5 +21,30 @@ void Device::memcpy(DevicePtr dst, DevicePtr src, uint64_t size) {
 
 #endif
 }
+
+void GraphicsDevice::image_transition(DeviceAllocation img,
+                                      ImageLayout old_layout,
+                                      ImageLayout new_layout) {
+  auto cmd_list = new_command_list({CommandListType::Graphics});
+  cmd_list->image_transition(img, old_layout, new_layout);
+  submit_synced(cmd_list.get());
+}
+void GraphicsDevice::buffer_to_image(DeviceAllocation dst_img,
+                                     DevicePtr src_buf,
+                                     ImageLayout img_layout,
+                                     const BufferImageCopyParams &params) {
+  auto cmd_list = new_command_list({CommandListType::Graphics});
+  cmd_list->buffer_to_image(dst_img, src_buf, img_layout, params);
+  submit_synced(cmd_list.get());
+}
+void GraphicsDevice::image_to_buffer(DevicePtr dst_buf,
+                                     DeviceAllocation src_img,
+                                     ImageLayout img_layout,
+                                     const BufferImageCopyParams &params) {
+  auto cmd_list = new_command_list({CommandListType::Graphics});
+  cmd_list->image_to_buffer(dst_buf, src_img, img_layout, params);
+  submit_synced(cmd_list.get());
+}
+
```

Comment:
These APIs will get new command list and submit it immediately. Will this cause too many command lists submitted in one frame?

Reply:
it will, but this is the "immediate mode" API. If possible we should avoid this and use the image_to_buffer from the command list directly. A potential issue with the current design is that each Taichi kernel dispatch is a unique command list. In TVM's design the commands are accumulated into one buffer and dispatched together upon a sync. Since we don't have a good sync boundary, the current design will start dispatching work to the GPU immediately but not necessarily wait for them to complete. That said, this is a choice by the taichi backend not the RHI, the RHI will support all cases.